anonlink/stats.py

```diff
@@ -0,0 +1,219 @@
+    that would be returned if the candidate_pairs were found using a
+    particular threshold. This function requires only a single pass of
+    the data, so it is faster than simply running the greedy solver
+    multiple times.
+
+    :param candidate_pairs: The candidate pairs.
+    :param steps: An integer determining the number of threshold steps
+        to use. Default 100.
+
+    :return: 2-tuple of (1) the number of matches for the threshold as
+        an array length steps + 1, and (2) the thresholds as an array
+        length steps + 1.
+    """
+    num_matches, _, thresholds = _semiopen_hist_matches_nonmatches(
+        candidate_pairs, steps)
+    num_matches_rev = num_matches[::-1]
+    _np.cumsum(num_matches_rev, out=num_matches_rev)
+    return num_matches, thresholds
+
+
+def plot_similarities_hist(candidate_pairs, bins=100):
+    """Plot a histogram of the similarity scores in candidate pairs.
+
+    This function is experimental and subject to change without warning.
+
+    Matplotlib is used as the plotting backend.
+
+    :param candidate_pairs: The candidate pairs.
+    :param bins: An integer determining the number of bins to use.
+        Default 100.
+    """
+    # Do not import Matplotlib until we know it's needed: importing it
+    # has annoying side effects.
+    import matplotlib.pyplot as plt
+
+    counts, bin_boundaries = similarities_hist(candidate_pairs, bins)
+    plt.hist(bin_boundaries[:-1], bins=bin_boundaries, weights=counts)
+    plt.yscale('symlog')
+    plt.ylabel('Count')
+    plt.xlabel('Similarity')
+    plt.title('Histogram of similarities')
+    plt.show()
+
+
+def plot_matches_nonmatches_ratio_hist(candidate_pairs, bins=100):
+    """Plot the ratio of possible matches and definite nonmatches.
+
+    This function is experimental and subject to change without warning.
+
+    We use the greedy solver to split the candidate pairs into possible
+    matches and definite nonmatches. A possible match may or may not be
+    accepted as a pair depending on the threshold chosen. A definite
+    nonmatch will never be accepted, since one record in this pair has
+    a more promising match with another record. We then plot the ratio
+    of possible matches and definite nonmatches.
+
+    Matplotlib is used as the plotting backend.
+
+    :param candidate_pairs: The candidate pairs.
+    :param bins: An integer determining the number of bins to use.
+        Default 100.
+    """
+    # Do not import Matplotlib until we know it's needed: importing it
+    # has annoying side effects.
+    import matplotlib.pyplot as plt
+
+    matches_num, nonmatches_num, bin_boundaries = matches_nonmatches_hist(
+        candidate_pairs, bins)
+    all_num = matches_num + nonmatches_num
+
+    nonmatches_ratio = [nm * 100 / (m + nm) if (m + nm) else 0
+                       for nm, m in zip(nonmatches_num, matches_num)]
+    matches_ratio = [m * 100 / (m + nm) if (m + nm) else 0
```

Comment:
Trivial but could be useful to refactor this out into a non-matplotlib function that is exposed.

I'm just imagining a javascript/html frontend where it might be nicer - not to mention slightly less revealing - if the Python library could compute the nonmatches/matches ratio too.

Reply:
Yes, that was my first thought too! I chose not to do it as some values are undefined (bins where matches + nonmatches = 0), so these need special handling. Options are:
- NaNs
- NumPy masked arrays

These are both fine options, but I thought it’d be neater to just return 2 arrays, and let the user perform the division and decide what to do with the undefined values. I don’t feel strongly about this though.

Would you have a preferred way of handling these values?

Reply:
Fair point - let's just leave it how you've done it